nt-be/src/handlers/balance_changes/counterparty.rs

```diff
@@ -0,0 +1,175 @@
+    let response: near_api::Data<FtMetadata> = contract
+        .call_function("ft_metadata", serde_json::json!({}))
+        .read_only()
+        .fetch_from(network)
+        .await?;
+
+    Ok(response.data)
+}
+
+/// Store or update FT token metadata in counterparties table
+pub async fn upsert_ft_counterparty(
+    pool: &PgPool,
+    account_id: &str,
+    metadata: &FtMetadata,
+) -> Result<(), Box<dyn std::error::Error>> {
+    sqlx::query!(
+        r#"
+        INSERT INTO counterparties (
+            account_id,
+            account_type,
+            token_symbol,
+            token_name,
+            token_decimals,
+            token_icon,
+            last_verified_at
+        ) VALUES ($1, 'ft_token', $2, $3, $4, $5, NOW())
+        ON CONFLICT (account_id) 
+        DO UPDATE SET
+            token_symbol = EXCLUDED.token_symbol,
+            token_name = EXCLUDED.token_name,
+            token_decimals = EXCLUDED.token_decimals,
+            token_icon = EXCLUDED.token_icon,
+            last_verified_at = NOW()
+        "#,
+        account_id,
+        metadata.symbol,
+        metadata.name,
+        metadata.decimals as i16,
+        metadata.icon,
+    )
+    .execute(pool)
+    .await?;
+
+    Ok(())
+}
+
+/// Get FT token decimals from counterparties table
+/// Returns None if not found or not an FT token
+pub async fn get_ft_decimals(
+    pool: &PgPool,
+    token_contract: &str,
+) -> Result<Option<u8>, Box<dyn std::error::Error>> {
+    let result = sqlx::query!(
+        r#"
+        SELECT token_decimals
+        FROM counterparties
+        WHERE account_id = $1 AND account_type = 'ft_token'
+        "#,
+        token_contract
+    )
+    .fetch_optional(pool)
+    .await?;
+
+    Ok(result.and_then(|r| r.token_decimals.map(|d| d as u8)))
+}
+
+/// Ensure FT token metadata exists in counterparties table
+/// If not found, queries the contract and stores it
+pub async fn ensure_ft_metadata(
+    pool: &PgPool,
+    network: &NetworkConfig,
+    token_contract: &str,
+) -> Result<u8, Box<dyn std::error::Error>> {
+    // Check if we already have the metadata
+    if let Some(decimals) = get_ft_decimals(pool, token_contract).await? {
+        return Ok(decimals);
+    }
+
+    // Query from contract and store
+    let metadata = query_ft_metadata(network, token_contract).await?;
+    let decimals = metadata.decimals;
+    upsert_ft_counterparty(pool, token_contract, &metadata).await?;
+
+    log::info!(
+        "Discovered FT token: {} ({}) with {} decimals",
+        metadata.name,
+        metadata.symbol,
+        decimals
+    );
+
+    Ok(decimals)
+}
+
+/// Convert raw FT amount to human-readable decimal string
+///
+/// # Arguments
+/// * `raw_amount` - The raw amount from ft_balance_of (smallest units)
+/// * `decimals` - Number of decimal places for this token
+///
+/// # Returns
+/// A decimal string like "2.5" instead of "2500000"
+pub fn convert_raw_to_decimal(
```

Comment:
Why do we store human-readable instead of a full token as a string? I feel that the precision might drive us crazy later on

Reply:
I was also considering back and forth about storing the raw amount or the decimal adjusted value, but came to the conclusion that it is better if the values in the database are ready to use for any calculation. Also the values can then be presented directly without the need to look up the token for decimal conversion.

When looking at it now, I think though I would return BigDecimal directly here, and not a string. So that refactoring I would like to do. Do you agree @akorchyn ?

Reply:
Maybe we should store both, so we could get both precise and ready for calculation value for USD too.

But not sure. I feel like having a precise value is less risky comparing to not having one. As for one we would need to reset and for other just migrate

On the other hand it's probably only be used in graph and recent activity, so small precision error is acceptable there.

This call is on you. Just wanted to highlight it

Reply:
It is stored as the `NUMERIC` type in postgres, so there should not be precision errors. I prefer keeping the decimal value then. But I would like to return BigDecimal directly from that conversion function, so I'll address that later. Thanks for pointing it out though. It's good to have a second opinion. Will also consider storing the raw value in addition, which could be a good reference.